### ISSUE TITLE
Linux headers include system headers

### ISSUE THREAD

**Issue by stevenengler** [shadow/shadow#3475]
Some linux headers include system headers. For example in `include/linux/if.h`:

```c
#include <sys/socket.h>         /* for struct sockaddr.     */
```

This can cause bindgen to fail where clang complains with "error: redefinition of _". We can avoid including headers in standard system directories by including the `-nostdlibinc` clang argument to bindgen, but then we get "error: 'sys/socket.h' file not found" errors instead.

An example of this is that we need to create bindings for `struct ifconf`, which is defined in `include/linux/if.h`. If we include "if.h" in our bindings, we get:

```text
src/lib/linux-api/bindings-build/linux-install/include/asm/signal.h:15:23: error: typedef redefinition with different types ('unsigned long' vs '__sigset_t')
src/lib/linux-api/bindings-build/linux-install/include/linux/time.h:16:8: error: redefinition of 'timeval'
./bindings-wrapper.h:121:5: error: redefinition of enumerator 'SHUT_RD'
./bindings-wrapper.h:122:5: error: redefinition of enumerator 'SHUT_WR'
./bindings-wrapper.h:123:5: error: redefinition of enumerator 'SHUT_RDWR'
```

For some types we could just copy the definition from the linux headers into `bindings-wrapper.h`. But this doesn't work for all types. For example `struct ifconf` uses `struct ifreq` which uses `struct sockaddr` from `sys/socket.h`. I'm not sure what the best solution here is. Maybe we should write our own `sys/*` headers which we make sure don't conflict with the linux headers. But this might be a lot of work and be easy to get wrong.

**Comment by sporksmith** [comment]
It seems surprising that linux includes libc headers, particularly since the main consumer of the linux headers is itself libc IIUC, but yeah here we are.

My best guess is that normally this isn't an issue since the libc headers that include `linux/if.h` don't also include a bunch of unrelated linux headers like `linux/signal.h`. Following that line of thinking we can sidestep the issue by splitting up our bindings generation to avoid these header incompatibilities. Proof of concept: #3476

**Comment by stevenengler** [comment]
> It seems surprising that linux includes libc headers, particularly since the main consumer of the linux headers is itself libc IIUC, but yeah here we are.

Yeah, this was a little frustrating to find. It's sort of funny that the following program fails to build:

```c
#include <linux/signal.h>
#include <linux/if.h>

int main() {
    return 0;
}
```

```text
In file included from test.c:2:
In file included from /usr/include/linux/if.h:28:
In file included from /usr/include/sys/socket.h:33:
In file included from /usr/include/bits/socket.h:29:
In file included from /usr/include/sys/types.h:179:
In file included from /usr/include/sys/select.h:33:
/usr/include/bits/types/sigset_t.h:7:20: error: typedef redefinition with different types ('__sigset_t' vs 'unsigned long')
    7 | typedef __sigset_t sigset_t;
      |                    ^
/usr/include/asm/signal.h:15:23: note: previous definition is here
   15 | typedef unsigned long sigset_t;
```

> My best guess is that normally this isn't an issue since the libc headers that include `linux/if.h` don't also include a bunch of unrelated linux headers like `linux/signal.h`. Following that line of thinking we can sidestep the issue by splitting up our bindings generation to avoid these header incompatibilities. Proof of concept: #3476

My concern with this approach is that we can end up with duplicate definitions in our bindings, and then we would have the same type defined in two different rust modules with no way to deduplicate them.

On the other hand, #3476 doesn't have this issue. So maybe it would be good to go with this approach until/if we do run into a case where we need to deal with duplicate definitions.

Another concern is that the bindings might become tied to whatever distro was used since it will include types from the system headers. I don't think this is a big deal, and we could just try to document in the commit message what distro the bindings were generated on, so that we know how to reproduce the exact same bindings in the future.

**Comment by sporksmith** [comment]
> My concern with this approach is that we can end up with duplicate definitions in our bindings, and then we would have the same type defined in two different rust modules with no way to deduplicate them.

Yeah, I think we can mostly address this by being a bit careful about what bindings we generate for such additional runs. In #3476 I iterated on this a bit in "squash" commits.

Currently it pulls in the `sockaddr` definition, rewrites it to `linux_sockaddr`, and exposes it in the binding since it's used by some of the other types being exposed. I *think* that's probably what we want in this case, since there is no `sockaddr` type in the linux headers for it to conflict with, and it doesn't hurt to further distinguish it from `libc::sockaddr`. In the higher level wrapper we expose we can add conversions between `linux_sockaddr` and `libc::sockaddr` when the libc feature is enabled.

I can imagine other cases where the type we're binding uses types that are already wrapped in the "main" bindgen run. In that case I think we *can* suppress emitting the duplicated type and instead put in a `use` line to get the definition we've already emitted. (I don't think we want to do this to inject `libc::sockaddr` above, since the dependency on the `libc` crate is optional).
